services/local-notifications/core.go

```diff
@@ -0,0 +1,322 @@
+	walletDB          *wallet.Database
+	accountsDB        *accounts.Database
+}
+
+func NewService(appDB *sql.DB, network uint64) *Service {
+	db := NewDB(appDB, network)
+	walletDB := wallet.NewDB(appDB, network)
+	accountsDB := accounts.NewDB(appDB)
+	trans := &transmitter{}
+	walletTrans := &transmitter{}
+
+	return &Service{
+		db:                db,
+		walletDB:          walletDB,
+		accountsDB:        accountsDB,
+		transmitter:       trans,
+		walletTransmitter: walletTrans,
+	}
+}
+
+func pushMessage(notification *Notification) {
+	log.Info("Pushing a new push notification", "info", notification)
+	signal.SendLocalNotifications(notification)
+}
+
+func (s *Service) buildTransactionNotification(rawTransfer wallet.Transfer) *Notification {
+	log.Info("Handled a new transfer in buildTransactionNotification", "info", rawTransfer)
+
+	var state = undetermined
+	transfer := wallet.CastToTransferView(rawTransfer)
+
+	switch {
+	case transfer.TxStatus == hexutil.Uint64(0):
+		state = failed
+	case transfer.Address == transfer.To:
+		state = inbound
+	default:
+		state = outbound
+	}
+
+	from, err := s.accountsDB.GetAccountByAddress(types.Address(transfer.From))
+
+	if err != nil {
+		log.Debug("Could not select From account by address", "error", err)
+	}
+
+	to, err := s.accountsDB.GetAccountByAddress(types.Address(transfer.To))
+
+	if err != nil {
+		log.Debug("Could not select To account by address", "error", err)
+	}
+
+	body := notificationBody{
+		State:       state,
+		From:        transfer.From,
+		To:          transfer.Address,
+		FromAccount: from,
+		ToAccount:   to,
+		Value:       transfer.Value,
+		ERC20:       string(transfer.Type) == "erc20",
+		Contract:    transfer.Contract,
+		Network:     transfer.NetworkID,
+	}
+
+	return &Notification{
+		ID:       transfer.ID,
+		Body:     body,
+		Category: "transaction",
+	}
+}
+
+func (s *Service) transactionsHandler(payload TransactionEvent) {
+	log.Info("Handled a new transaction", "info", payload)
+
+	limit := 20
+	if payload.BlockNumber != nil {
+		for _, address := range payload.Accounts {
+			log.Info("Handled transfer for address", "info", address)
+			transfers, err := s.walletDB.GetTransfersByAddressAndBlock(address, payload.BlockNumber, int64(limit))
+			if err != nil {
+				log.Error("Could not fetch transfers", "error", err)
+			}
+
+			for _, transaction := range transfers {
+				n := s.buildTransactionNotification(transaction)
+				pushMessage(n)
+			}
+		}
+	}
+}
+
+// SubscribeWallet - Subscribes to wallet signals
+func (s *Service) SubscribeWallet(publisher *event.Feed) error {
+	s.walletTransmitter.publisher = publisher
+
+	preference, err := s.db.GetWalletPreference()
+
+	if err != nil {
+		log.Error("Failed to get wallet preference", "error", err)
+		return nil
+	}
+
+	if preference.Enabled {
+		s.StartWalletWatcher()
+	}
+
+	return nil
+}
+
+// StartWalletWatcher - Forward wallet events to notifications
+func (s *Service) StartWalletWatcher() {
```

Comment:
Is the approach used here to subscribe to the wallet correct in terms of code style?

Reply:
Nice early returns, but I do have a question